beebop/utils.py

```diff
@@ -82,9 +83,82 @@ def replace_filehashes(folder: str, filename_dict: dict) -> None:
             print(line)
 
 
-def add_query_ref_status(
-    fs: PoppunkFileStore, p_hash: str, filename_dict: dict
-) -> None:
+def create_subgraphs(network_folder: str, filename_dict: dict) -> None:
+    """
+    [Create subgraphs for the network visualisation. These are what
+    will be sent back to the user to see.
+    The subgraphs are created
+    by selecting a maximum number nodes, prioritizing query nodes and adding
+    neighbor nodes until the maximum number of nodes is reached. The query
+    nodes are highlighted in the network graph by adding a ref or query status
+    to the .graphml files.]
+
+    :param network_folder: [path to the network folder]
+    :param filename_dict: [dict that maps filehashes(keys) toclear
+        corresponding filenames (values) of all query samples. We only need
+        the filenames here.]
+    """
+    query_names = list(filename_dict.values())
+
+    for path in get_component_filenames(network_folder):
+        SubGraph = build_subgraph(path, query_names)
+
+        add_query_ref_to_graph(SubGraph, query_names)
+
+        write_graphml(
+            SubGraph,
+            path.replace("network_component", "pruned_network_component"),
+        )
+
+
+def get_component_filenames(network_folder: str) -> list[str]:
+    """
+    [Get all network component filenames in the network folder.]
+
+    :param network_folder: [path to the network folder]
+    :return list: [list of all network component filenames]
+    """
+    return glob.glob(network_folder + "/network_component_*.graphml")
+
+
+def build_subgraph(path: str, query_names: list) -> Graph:
+    """
+    [Build a subgraph from a network graph, prioritizing query nodes and
+    adding neighbor nodes until the maximum number of nodes is reached.]
+
+    :param path: [path to the network graph]
+    :param query_names: [list of query sample names]
+    :return nx.Graph: [subgraph]
+    """
+    MAX_NODES = 30  # arbitrary number based on performance
+    Graph = read_graphml(path)
+
+    # get query nodes
+    query_nodes = {
+        node for (node, id) in Graph.nodes(data="id") if id in query_names
+    }
+
+    # get neighbor nodes of query nodes
+    neighbor_nodes = set()
+    for node in query_nodes:
+        neighbor_nodes.update(Graph.neighbors(node))
+
+    # remove query nodes from neighbor nodes
+    neighbor_nodes = neighbor_nodes - query_nodes
+
+    # create final set of nodes, prioritizing query nodes
+    sub_graph_nodes = set()
+    sub_graph_nodes.update(query_nodes)
+
+    # add neighbor nodes until we reach the maximum number of nodes
+    remaining_capacity = MAX_NODES - len(sub_graph_nodes)
+    if remaining_capacity > 0:
+        sub_graph_nodes.update(list(neighbor_nodes)[:remaining_capacity])
+
+    return Graph.subgraph(sub_graph_nodes)
```

Comment:
My only concern about this logic is the case where you may have multiple query nodes in a cluster - as we discussed, these will all be included in the subgraph.. but you can imagine a situation where two samples came from widely different parts of the cluster, and all of the neighbour node capacity was taken up by the neighbours of the first sample - which would then make the second sample appear completely isolated.

Maybe in those cases each sample in the cluster should get MAX_NODES capacity for its own neighbours...?